src/AspNet.Security.OAuth.Shopify/ShopifyAuthenticationHandler.cs

```diff
@@ -0,0 +1,218 @@
+using System.Linq;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Security.Claims;
+using System.Text.Encodings.Web;
+using System.Threading.Tasks;
+using JetBrains.Annotations;
+using Microsoft.AspNetCore.Authentication;
+using Microsoft.AspNetCore.Authentication.OAuth;
+using Microsoft.AspNetCore.WebUtilities;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace AspNet.Security.OAuth.Shopify
+{
+    /// <inheritdoc />
+    [UsedImplicitly]
+    public class ShopifyAuthenticationHandler : OAuthHandler<ShopifyAuthenticationOptions>
+    {
+        /// <inheritdoc />
+        public ShopifyAuthenticationHandler(
+            [NotNull] IOptionsMonitor<ShopifyAuthenticationOptions> options, 
+            [NotNull] ILoggerFactory logger, 
+            [NotNull] UrlEncoder encoder, 
+            [NotNull] ISystemClock clock) 
+            : base(options, logger, encoder, clock)
+        {
+        }
+
+        /// <inheritdoc />
+        protected override async Task<AuthenticationTicket> CreateTicketAsync(
+            [NotNull] ClaimsIdentity identity,
+            [NotNull] AuthenticationProperties properties, 
+            [NotNull] OAuthTokenResponse tokens)
+        {
+            var uri = string.Format(ShopifyAuthenticationDefaults.UserInformationEndpoint,
+                properties.Items[ShopifyAuthenticationDefaults.ShopNameAuthenticationProperty]);
+
+            var request = new HttpRequestMessage(HttpMethod.Get, uri);
+            request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+            request.Headers.Add("X-Shopify-Access-Token", tokens.AccessToken);
+
+            var response = await Backchannel.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, Context.RequestAborted);
+            if (!response.IsSuccessStatusCode)
+            {
+                Logger.LogError("An error occurred while retrieving the user profile: the remote server " +
+                                "returned a {Status} response with the following payload: {Headers} {Body}.",
+                                /* Status: */ response.StatusCode,
+                                /* Headers: */ response.Headers.ToString(),
+                                /* Body: */ await response.Content.ReadAsStringAsync());
+            
+                throw new HttpRequestException("An error occurred while retrieving the shop profile.");
+            }
+
+            var payload = JObject.Parse(await response.Content.ReadAsStringAsync());
+
+#if DEBUG
+            // ReSharper disable once UnusedVariable
+            var jsonStr = payload.ToString(Formatting.Indented);
+#endif
+
+            // In Shopify, the customer can modify the scope given to the app. Apps should verify
+            // that the customer is allowing the required scope.
+            var actualScope = tokens.Response["scope"].ToString();
```

Comment:
Does this need to access `scope` in a safer way (e.g. if it's not there or is null)?